src/docfx/lib/markdown/validation/DocsValidationExtension.cs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Markdig;
+using Markdig.Syntax;
+using Markdig.Syntax.Inlines;
+using Microsoft.DocAsCode.MarkdigEngine.Extensions;
+using Microsoft.Docs.Validation;
+
+namespace Microsoft.Docs.Build
+{
+    internal static class DocsValidationExtension
+    {
+        public static MarkdownPipelineBuilder UseDocsValidation(
+            this MarkdownPipelineBuilder builder,
+            MarkdownEngine markdownEngine,
+            ContentValidator contentValidator)
+        {
+            return builder.Use(document =>
+            {
+                var file = (Document)InclusionContext.File;
+                if (file.FilePath.Format != FileFormat.Markdown)
+                {
+                    return;
+                }
+
+                var documentNodes = new List<ContentNode>();
+                var inclusionDocumentNodes = new Dictionary<Document, List<ContentNode>>();
+                MarkdigUtility.Visit(document, null, null, (node, parents, current) =>
+                {
+                    if (node is HeadingBlock headingBlock)
+                    {
+                        var heading = new Heading
+                        {
+                            Level = headingBlock.Level,
+                            SourceInfo = headingBlock.GetSourceInfo(),
+                            ParentSourceInfoList = parents?.Cast<object?>().ToList() ?? new List<object?>(),
+                            Content = GetHeadingContent(headingBlock), // used for reporting
+                            HeadingChar = headingBlock.HeaderChar,
+                            RenderedPlainText = markdownEngine.ToPlainText(headingBlock), // used for validation
+                            IsVisible = MarkdigUtility.IsVisible(headingBlock),
+                        };
+
+                        if (current != null && parents != null)
```

Comment:
Consider use a `bool` to represent `IsIncIude`. I might expect `current` to be `InclusionContext.File` for none-inclusion nodes.